Unconditionally define POSIX sysconf macros PAGE_SIZE and PAGESIZE

besmarsh opened the pull request:
~The POSIX `sysconf()` option `_SC_PAGESIZE` was defined as `PAGESIZE`, which was defined as `PAGE_SIZE` (coming from `CONFIG_POSIX_PAGE_SIZE`) IF not already defined. The STM32 HAL defines `PAGESIZE` for flash pages as `FLASH_PAGE_SIZE`, which is only defined in the HAL for some STM32 families. This caused build failures for many STM32 targets. See issue #74623 for full description.~

~In addition, the POSIX `sysconf()` options `_SC_PAGE_SIZE` and `_SC_PAGESIZE` should be synonymous, but the existing definition could result in them differing when `PAGESIZE` is already defined elsewhere.~

~This issue was present in v3.6 but is more of an issue for v3.7 as it causes build failures for applications not even using sysconf at all, if they target an affected STM32 family and enable `CONFIG_POSIX_API` and `CONFIG_CPP`.~

~This PR removes the conditional definition of `PAGESIZE` and uses `PAGE_SIZE` (set from `CONFIG_POSIX_PAGE_SIZE`) for both `_SC_PAGE_SIZE` and `_SC_PAGESIZE`.~

The POSIX macros `PAGE_SIZE` and `PAGESIZE` (queriable through `sysconf()`) were conditionally defined only if an existing definition did not already exist. These should be defined unconditionally in their header to ensure they get the correct values.

If these macros are defined elsewhere with a different meaning, that's a problem. There was an issue where `PAGESIZE` was already defined with a different meaning. See https://github.com/zephyrproject-rtos/zephyr/issues/74623 and https://github.com/zephyrproject-rtos/zephyr/pull/74428.

The POSIX macro `ATEXIT_MAX` is also conditionally defined and should be unconditionally defined, but there is currently a definition in picolibc (`picolibc/newlib/libc/include/stdlib.h`) so this change will be done separately.

This PR defines `PAGE_SIZE` and `PAGESIZE` unconditionally.

besmarsh committed:
posix: sys: sysconf: Unconditionally define POSIX macros

The POSIX macros PAGE_SIZE and PAGESIZE (queriable through
sysconf()) were conditionally defined only if an existing definition
did not already exist. These should be defined unconditionally in their
header to ensure they get the correct values.

If these macros are defined elsewhere with a different meaning, that's a
problem. There was an issue where PAGESIZE was already defined with a
different meaning. See #74623 and #74428.

The POSIX macro ATEXIT_MAX is also conditionally defined and should be
unconditionally defined, but there is currently a definition in picolibc
(picolibc/newlib/libc/include/stdlib.h) so this change will be done
separately.

This commit defines PAGE_SIZE and PAGESIZE unconditionally.

Signed-off-by: Ben Marsh <ben.marsh@helvar.com>

besmarsh commented:
I've just seen #70136, #73363, and #74428 regarding this same issue. The approach there was to make changes to the STM32 HAL - I guess that solves the issue too, but I think the fix here is good for making the `_SC_PAGE_SIZE` and `_SC_PAGESIZE` options consistent and equal.

cfriedt commented:
#74428 should fix this

besmarsh commented:
This is fixed by #74428, but this change ensures that no other rogue `PAGESIZE` definitions can cause the same issue, and that `_SC_PAGE_SIZE` and `_SC_PAGESIZE` always return the same.

cfriedt commented on include/zephyr/posix/posix_features.h:
```diff
@@ -318,10 +318,6 @@
 #define PAGE_SIZE CONFIG_POSIX_PAGE_SIZE
 #endif
 
-#ifndef PAGESIZE
-#define PAGESIZE PAGE_SIZE
-#endif
-
```
```suggestion
#define PAGESIZE PAGE_SIZE
```

besmarsh replied:
Should the conditions around the definitions of `PAGE_SIZE` and `ATEXIT_MAX` just above also be removed to ensure they are defined correctly?

cfriedt commented on include/zephyr/posix/sys/sysconf.h:
```diff
@@ -278,7 +278,7 @@ enum {
 #define __z_posix_sysconf_SC_OPEN_MAX                     CONFIG_ZVFS_OPEN_MAX
 #define __z_posix_sysconf_SC_PAGE_SIZE                    PAGE_SIZE
-#define __z_posix_sysconf_SC_PAGESIZE                     PAGESIZE
+#define __z_posix_sysconf_SC_PAGESIZE                     PAGE_SIZE
```
It would be preferable to keep the original definition here. Otherwise, it would be an unnecessary special case for scripts.

besmarsh replied:
Keeping the original definition is fine if `PAGESIZE` and `PAGE_SIZE` are unconditionally defined equal.

cfriedt replied:
Yes

cfriedt reviewed (commented):
Removing the conditions around defining `PAGESIZE` should be all that is needed here, but `PAGESIZE` also needs to be defined according to the spec.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html

besmarsh commented:
PR updated, including change of title and description after scope change.

besmarsh commented:
`ATEXIT_MAX` is also defined in `picolibc/newlib/libc/include/stdlib.h` - I don't know if it should be there or not

cfriedt commented:
@besmarsh - if you prefer, you can go back to the way it was before.

There will be a "Great Reckoning" not far down the road where we make internal Zephyr headers as consistent as possible with both Picolibc and Newlib.

erwango commented:
Removing the `STM32` label as this part of the issues has been solved

besmarsh commented:
> @besmarsh - if you prefer, you can go back to the way it was before.
> 
> There will be a "Great Reckoning" not far down the road where we make internal Zephyr headers as consistent as possible with both Picolibc and Newlib.

I've removed the `ATEXIT_MAX` change and just changed `PAGE_SIZE` and `PAGESIZE`. The `ATEXIT_MAX` change can wait for the alignment of Zephyr and Picolibc/Newlib headers.
